src/D3MHub.sol

```diff
@@ -221,121 +219,118 @@ contract D3MHub {
-                                availableAssets
-                            ),
-                            daiDebt
-                        );
-        require(amount <= MAXINT256, "D3MHub/overflow");
+    function _unwind(bytes32 ilk, ID3MPool _pool, uint256 amount) internal {
+        if (amount > 0) {
+            _pool.withdraw(amount);
+            daiJoin.join(address(this), amount);
+            vat.frob(ilk, address(_pool), address(_pool), address(this), -int256(amount), -int256(amount));
+            vat.slip(ilk, address(_pool), -int256(amount));
+        }
+        emit Unwind(ilk, amount);
+    }
 
-        // Determine the amount of fees to bring back
-        uint256 fees = 0;
-        if (assetBalance > daiDebt) {
-            unchecked {
-                fees = assetBalance - daiDebt;
-            }
+    function _unwind(bytes32 ilk, ID3MPool _pool, address urn) internal {
+        uint256 amount = _pool.maxWithdraw();
+        if (amount > 0) {
+            uint256 toSlip = _min(vat.gem(ilk, urn), amount);
+            require(toSlip <= MAXINT256, "D3MHub/overflow");
+            vat.slip(ilk, urn, -int256(toSlip));
 
-            if (amount + fees > availableAssets) {
-                unchecked {
-                    fees = availableAssets - amount;
-                }
-            }
+            _pool.withdraw(amount);
+            daiJoin.join(address(this), amount);
+            vat.move(address(this), vow, amount * RAY);
         }
+        emit Unwind(ilk, amount);
+    }
 
-        if (amount == 0 && fees == 0) {
-            emit Unwind(ilk, 0, 0);
-            return;
+    function _normal(bytes32 ilk, ID3MPool _pool, uint256 Art, uint256 lineWad) internal {
+        (uint256 ink, uint256 art) = vat.urns(ilk, address(_pool));
+        require(art == Art, "D3MHub/more-than-one-urn");
+        uint256 currentAssets = _pool.assetBalance(); // Should return DAI owned by D3MPool
+        uint256 maxWithdraw = _pool.maxWithdraw();
+        if (currentAssets > ink) { // If fees were generated
+            uint256 fixInk = _min(
+                currentAssets - ink, // fees generated
+                ink < lineWad // if previously CDP was under debt ceiling
+                    ? (lineWad - ink) + maxWithdraw // up to gap to reach debt ceiling + maxWithdraw
+                    : maxWithdraw // up to maxWithdraw
+            );
+            require(fixInk <= MAXINT256, "D3MHub/overflow");
+            vat.slip(ilk, address(_pool), int256(fixInk)); // Generate extra collateral
+            vat.frob(ilk, address(_pool), address(_pool), address(this), int256(fixInk), 0); // Lock it
+            ink += fixInk;
+            emit Fees(ilk, fixInk);
+        }
+        if (art < ink) { // If there was permissionless DAI paid or fees added as collateral
+            address _vow = vow;
+            uint256 fixArt = ink - art; // Amount of fees + permissionless DAI paid we will now transform to debt
```

Comment:
can be `unchecked`

Reply:
> Just noticed that in `cull()` the call gets blocked if either `ink` or `art` are over `MAXINT256` which could potentially be a valid state in the extreme. We could also change this to take the `_min()` of each value and the `MAXINT256` so that we can at least partially `grab()`. This would require one more `vat.urn()` check after to ensure we collected all of the `ink` and `art` before setting `ilks[ilk].culled` to 1 though.
> 
> Not sure what the implications of a partial `cull()` and then a call to `exec()` would be though, so that would require some investigation.

As Kurt said, I think that would be an extreme case that would break other stuff break, and to be honest if it is `cull` breaking, that is something governance can fix. Partial grabbing can bring more unintended consequences.

Reply:
Done!